src/monomorphize.rs

```diff
@@ -11,3 +11,12 @@ pub fn apply_param_substs<'a, 'tcx,T>(tcx: &TyCtxt<'a, 'tcx, 'tcx>,
     let substituted = value.subst(*tcx, param_substs);
     tcx.normalize_associated_type(&substituted)
 }
+
+pub fn apply_ty_substs<'a, 'tcx>(tcx: &TyCtxt<'a, 'tcx, 'tcx>,
```

Comment:
This part would still be necessary, but it should be in a common place. Merging miri into `librustc_mir` would be a good way to get that.